docs/lang/articles/advanced/layout.md

```diff
@@ -3,58 +3,37 @@ sidebar_position: 2
+There arises two questions in making performance optimization efforts: 1) how to organize a faster data layout and 2) how to manage memory occupancy. In this article, we discuss in the scope of Taichi field. A better understanding of the mechanism under the hood is essential to write blazing fast Taichi programs.
 
-## Packed mode
+If you feel confused about the basic concepts of Taichi, please refer to the basic document [Fields (basic)](../basic/field).
 
-By default, all non-power-of-two dimensions of a field are automatically
-padded to a power of two. For instance, a field of shape `(18, 65)` will
-have an internal shape of `(32, 128)`. Although the padding has many benefits
-such as allowing fast and convenient bitwise operations for coordinate handling,
-it will consume potentially much more memory than expected.
 
-If you would like to reduce memory usage, you can use the optional packed
-mode. In packed mode, no padding is applied such that a field does not have a
-larger internal shape than the defined shape when some of its dimensions
-are not powers of two. The downside is that the runtime performance will
-regress slightly.
+## How to organize an efficient data layout
 
-A switch named `packed` for `ti.init()` decides whether to use packed mode:
+In this section, we introduce how to organize data layouts in Taichi fields. The central principle of efficient data layout is locality. Generally speaking, a program with desirable locality has at least one of the following features:
+* Dense data structure
+* Loop over data in small-range (range within 32KB is especially good for most processors)
+* Sequential load/store
 
-```python
-ti.init()  # default: packed=False
-a = ti.field(ti.i32, shape=(18, 65))  # padded to (32, 128)
-```
-
-```python
-ti.init(packed=True)
-a = ti.field(ti.i32, shape=(18, 65))  # no padding
-```
-
-## Advanced data layouts
+:::note
 
-Apart from shape and data type, you can also specify the data layout of a
-field in a recursive manner. This may allow you to achieve better performance.
+Brief explanations:
 
-Normally, you don't have to worry about the performance nuances between
-different layouts, and you can just use the default one (simply by specifying
-`shape` when creating fields) as a start.
+Be aware that data are always fetched from memory in blocks (pages). The hardware has no knowledge about the usefulness of one specific data element in the block. The processor blindly fetch the entire block which _contains_ the requested memory address. Therefore, the memory bandwidth is wasted when data in the block are not well utilized.
 
-However, when a field gets large, a proper data layout may be critical to
-performance, especially for memory-bound applications. A carefully designed
-data layout has much better spatial locality, which will significantly
-improve cache/TLB-hit rates and cache line utilization.
+If sparsity is inevitable, refer to [Sparse computation](./sparse.md).
```

Comment:
Just to not sounds like we discourage the usage of sparse data structure. ;/
```suggestion
For sparse fields, refer to [Sparse computation](./sparse.md).
```

Reply:
Yeah, positive tone is important, thanks!